pkg/parts/parts.go

```diff
@@ -1,106 +1,164 @@
 package parts
 
 import (
-	"bytes"
 	"context"
 	"encoding/json"
 	"fmt"
-	"github.com/virtual-vgo/vvgo/pkg/locker"
+	"github.com/virtual-vgo/vvgo/pkg/log"
 	"github.com/virtual-vgo/vvgo/pkg/projects"
-	"github.com/virtual-vgo/vvgo/pkg/storage"
+	"github.com/virtual-vgo/vvgo/pkg/redis"
+	"github.com/virtual-vgo/vvgo/pkg/tracing"
+	"strconv"
 	"strings"
+	"sync"
 	"time"
 )
 
-const DataFile = "parts.json"
+var logger = log.Logger()
 
 var (
 	ErrInvalidPartName   = fmt.Errorf("invalid part name")
 	ErrInvalidPartNumber = fmt.Errorf("invalid part number")
 )
 
-type Parts struct {
-	*storage.Cache
-	*locker.Locker
+type RedisParts struct {
+	namespace string
 }
 
-func (x Parts) Init(ctx context.Context) error {
-	return x.PutObject(ctx, DataFile, storage.NewJSONObject([]byte(`[]`)))
+func NewParts(namespace string) *RedisParts {
+	return &RedisParts{namespace: namespace}
 }
 
-func (x Parts) List(ctx context.Context) ([]Part, error) {
-	// grab the data file
-	var data storage.Object
-	if err := x.GetObject(ctx, DataFile, &data); err != nil {
+// List returns a slice of all parts in the database.
+// If the first request to redis fails, this function will return an error.
+// Subsequent errors will only be logged.
+func (x *RedisParts) List(ctx context.Context) ([]Part, error) {
+	localCtx, span := tracing.StartSpan(ctx, "RedisParts.List()")
+	defer span.Send()
+
+	// Read the all the part _keys_ first.
+	partKeys, err := x.readIndex(localCtx)
+	if err != nil {
 		return nil, err
 	}
 
-	// deserialize the data file
-	var parts []Part
-	if err := json.NewDecoder(bytes.NewReader(data.Bytes)).Decode(&parts); err != nil {
-		return nil, fmt.Errorf("json.Decode() failed: %v", err)
+	// Now we can read each individual part data.
+	// Radix will automatically batch/pipeline requests for us, so we can queue up a bunch of requests in goroutines.
+	// https://pkg.go.dev/github.com/mediocregopher/radix/v3?tab=doc#hdr-Implicit_pipelining
+	parts := make([]Part, len(partKeys))
+	var wg sync.WaitGroup
+	wg.Add(2 * len(partKeys))
+	for i := range partKeys {
+		parts[i].DecodeRedisKey(partKeys[i])
+		go func(key string, dest *Part) {
+			sheetsKey := x.namespace + ":parts:" + key + ":sheets"
+			dest.Sheets = readLinks(ctx, sheetsKey)
+			wg.Done()
+		}(partKeys[i], &parts[i])
+
+		go func(key string, dest *Part) {
+			clixKey := x.namespace + ":parts:" + key + ":clix"
+			dest.Clix = readLinks(ctx, clixKey)
+			wg.Done()
+		}(partKeys[i], &parts[i])
 	}
+	wg.Wait()
 	return parts, nil
 }
 
-func (x Parts) Save(ctx context.Context, parts []Part) error {
-	// first, validate all the parts
-	if err := validatePart(parts); err != nil {
-		return err
+func (x *RedisParts) readIndex(localCtx context.Context) ([]string, error) {
+	var partKeys []string
+	if err := redis.Do(localCtx, redis.Cmd(&partKeys, "ZRANGE", x.namespace+":parts:index", "0", "-1")); err != nil {
+		return nil, err
 	}
+	return partKeys, nil
+}
```

Comment:
Why not shard the parts index by project? This reduces the growth factor of this lookup.

Reply:
This is a super cheap lookup, like kilobytes of data. We don't need to shard.

Reply:
Additionally, these are scored, so I'm not sure what additional advantage we get from sharding.